pkg/services/authn/authnimpl/sync/user_sync.go

```diff
@@ -63,29 +80,88 @@ var (
+		isUserProvisioningEnabled:  scimSection.Key("user_sync_enabled").MustBool(false),
+		userService:                userService,
+		authInfoService:            authInfoService,
+		userProtectionService:      userProtectionService,
+		quotaService:               quotaService,
+		log:                        log.New("user.sync"),
+		tracer:                     tracer,
+		features:                   features,
+		lastSeenSF:                 &singleflight.Group{},
 	}
 }
 
 type UserSync struct {
-	userService           user.Service
-	authInfoService       login.AuthInfoService
-	userProtectionService login.UserProtectionService
-	quotaService          quota.Service
-	log                   log.Logger
-	tracer                tracing.Tracer
-	features              featuremgmt.FeatureToggles
-	lastSeenSF            *singleflight.Group
+	allowedNonProvisionedUsers bool
+	isUserProvisioningEnabled  bool
+	userService                user.Service
+	authInfoService            login.AuthInfoService
+	userProtectionService      login.UserProtectionService
+	quotaService               quota.Service
+	log                        log.Logger
+	tracer                     tracing.Tracer
+	features                   featuremgmt.FeatureToggles
+	lastSeenSF                 *singleflight.Group
+}
+
+// ValidateUserProvisioningHook validates if a user should be allowed access based on provisioning status and configuration
+func (s *UserSync) ValidateUserProvisioningHook(ctx context.Context, id *authn.Identity, _ *authn.Request) error {
+	s.log.FromContext(ctx).Debug("Validating user provisioning", "auth_module", id.AuthenticatedBy, "auth_id", id.AuthID)
+	ctx, span := s.tracer.Start(ctx, "user.sync.ValidateUserProvisioningHook")
+	defer span.End()
+
+	// Skip validation if user provisioning is disabled
+	if !s.isUserProvisioningEnabled {
+		s.log.FromContext(ctx).Debug("User provisioning is disabled, skipping validation")
+		return nil
+	}
+
+	// Skip validation if non-provisioned users are allowed
+	if s.allowedNonProvisionedUsers {
+		s.log.FromContext(ctx).Debug("User provisioning is enabled, but non-provisioned users are allowed, skipping validation")
+		return nil
+	}
+
+	// Skip validation if the auth module is GrafanaComAuthModule
+	if id.AuthenticatedBy == login.GrafanaComAuthModule {
+		s.log.FromContext(ctx).Debug("User is authenticated via GrafanaComAuthModule, skipping validation")
+		return nil
+	}
+
+	// In order to guarantee the provisioned user is the same as the identity,
+	// we must validate the authinfo.ExternalUID with the identity.ExternalUID
+
+	// Retrieve user and authinfo from database
+	usr, authInfo, err := s.getUser(ctx, id)
+	if err != nil {
+		s.log.FromContext(ctx).Error("Failed to fetch user for validation", "error", err, "auth_module", id.AuthenticatedBy, "auth_id", id.AuthID)
+		return errUnableToRetrieveUserOrAuthInfo.Errorf("unable to retrieve user or authInfo for validation")
+	}
+
+	if usr == nil {
```

Comment:
Can this happen?

Reply:
In a super rare case of a race condition, but it is possible

WDYT? Keep it or not?

Reply:
> In a super rare case of a race condition, but it is possible

Then keep it!